internal/server/singleprocess/state/instance_exec.go

```diff
@@ -98,6 +98,65 @@ func (s *State) InstanceExecCreateByTargetedInstance(id string, exec *InstanceEx
 	return nil
 }
 
+// InstanceExecCreateForVirtualInstance registers the given InstanceExec record on
+// the instance specified. The instance does not yet have to be known (as it may
+// not yet have connected to the server) so this code will use memdb watchers
+// to detect the instance as it connects and then register the exec.
+func (s *State) InstanceExecCreateForVirtualInstance(ctx context.Context, id string, exec *InstanceExec) error {
+	// If the caller specified an instance id already, then just validate it.
+	if id == "" {
+		return status.Errorf(codes.NotFound, "No instance id given")
+	}
+
+	for {
+		// We have to start a new txn per iteration because we want to be able to observe
+		// the newly created record for the instance.
+		txn := s.inmem.Txn(false)
```

Comment:
I believe you forget to `txn.Abort` clean up this.

Reply:
I looked and abort is a noop for read txns so I just didn't bother, since I didn't want to accumulate defer's on the stack.

Reply:
Ah, can you note that in the comments please?